packages/compiler/core/projector.ts

```diff
@@ -0,0 +1,538 @@
+    const projectedType = projectType(variant.type);
+    const projectedDecs = projectDecorators(variant.decorators);
+
+    const projectedVariant = shallowClone(variant, {
+      type: projectedType,
+      decorators: projectedDecs,
+    });
+
+    checker.finishType(projectedVariant);
+    return projectedVariant;
+  }
+
+  function projectArray(array: ArrayType) {
+    const projectedType = projectType(array.elementType);
+
+    const projectedArray = shallowClone(array, {
+      elementType: projectedType,
+    });
+
+    checker.finishType(projectedArray);
+    return projectedArray;
+  }
+
+  function projectTuple(tuple: TupleType) {
+    const values: Type[] = [];
+    const projectedTuple = shallowClone(tuple, {
+      values,
+    });
+
+    for (const item of tuple.values) {
+      values.push(projectType(item));
+    }
+
+    return projectedTuple;
+  }
+
+  function projectEnum(e: EnumType) {
+    const members: EnumMemberType[] = [];
+    const decorators = projectDecorators(e.decorators);
+    const projectedEnum = shallowClone(e, {
+      members,
+      decorators,
+    });
+
+    for (const member of e.members) {
+      const projectedMember = projectType(member);
+      if (projectedMember.kind === "EnumMember") {
+        members.push(projectedMember);
+      }
+    }
+
+    checker.finishType(projectedEnum);
+    return applyProjection(e, projectedEnum);
+  }
+
+  function projectEnumMember(e: EnumMemberType) {
+    const decorators = projectDecorators(e.decorators);
+    const projectedMember = shallowClone(e, {
+      decorators,
+      enum: projectedTypes.get(e.enum)! as EnumType,
+    });
+
+    checker.finishType(projectedMember);
+    return projectedMember;
+  }
+
+  function projectDecorators(decs: DecoratorApplication[]) {
+    const decorators: DecoratorApplication[] = [];
+    for (const dec of decs) {
+      const args = [];
+      for (const arg of dec.args) {
+        // filter out primitive arguments
+        if (typeof arg !== "object") {
+          args.push(arg);
+          continue;
+        }
+
+        const projected = projectType(arg);
+        args.push(projected);
+      }
+
+      decorators.push({ ...dec, args });
+    }
+
+    return decorators;
+  }
+
+  function getInScopeProjections() {
+    const candidates = new Set(projections);
+    const inScope: ProjectionApplication[] = [];
+    let currentScope = namespaceScope();
+
+    outer: while (currentScope) {
+      for (const candidate of candidates) {
+        if (!candidate.scope || candidate.scope === currentScope) {
+          candidates.delete(candidate);
+          inScope.push(candidate);
+
+          if (candidates.size === 0) {
+            break outer;
+          }
+        }
+      }
+      currentScope = currentScope.namespace;
+    }
+
+    return inScope;
+  }
+
+  function namespaceScope(): NamespaceType | undefined {
+    for (let i = scope.length - 1; i >= 0; i--) {
+      if ((scope[i] as any).namespace !== undefined) {
```

Comment:
`if ("namespace" in scope[i]) {`
same as interface below? With the newer typescript version I was under the impression this would save us from having to cast ot any after too(we need to update I think)

Reply:
I don't understand the question here, can you elaborate?

Reply:
Just saying you could replace that condition with `if ("namespace" in scope[i])` like it is done a few lines below in this file.

I think typescript is making improvements to that style of check so you can reduce the use of `any`
```ts
if("bar" in foo) {
   foo.bar
}
```

Currently 
![image](https://user-images.githubusercontent.com/1031227/152030956-35bd2d0f-5300-429f-9955-895660aa372b.png)
but I was under the impression that an incoming change in typescript was solving this issue

Reply:
Ahh yeah, I think that should work in latest but doesn't work now. Although even with latest I find the guard doesn't work for me, and I can't say why.